jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/rabit/RabitTracker.scala

```diff
@@ -0,0 +1,193 @@
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+package ml.dmlc.xgboost4j.scala.rabit
+
+import java.net.{InetAddress, InetSocketAddress}
+
+import akka.actor.ActorSystem
+import akka.pattern.ask
+import ml.dmlc.xgboost4j.java.IRabitTracker
+import ml.dmlc.xgboost4j.scala.rabit.handler.RabitTrackerHandler
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future, Promise}
+import scala.util.{Failure, Success, Try}
+
+/**
+  * Scala implementation of the Rabit tracker interface without Python dependency.
+  * The Scala Rabit tracker fully implements the timeout logic, effectively preventing the tracker
+  * (and thus any distributed tasks) to hang indefinitely due to network issues or worker node
+  * failures.
+  *
+  * Note that this implementation is currently experimental, and should be used at your own risk.
+  *
+  * Example usage:
+  * {{{
+  *   import scala.concurrent.duration._
+  *
+  *   val tracker = new RabitTracker(32)
+  *   // allow up to 10 minutes for all workers to connect to the tracker.
+  *   tracker.start(10 minutes)
+  *
+  *   /* ...
+  *      launching workers in parallel
+  *      ...
+  *   */
+  *
+  *   // wait for worker execution up to 6 hours.
+  *   // providing a finite timeout prevents a long-running task from hanging forever in
+  *   // catastrophic events, like the loss of an executor during model training.
+  *   tracker.waitFor(6 hours)
+  * }}}
+  *
+  * @param numWorkers Number of distributed workers from which the tracker expects connections.
+  * @param port The minimum port number that the tracker binds to.
+  *             If port is omitted, or given as None, a random ephemeral port is chosen at runtime.
+  * @param maxPortTrials The maximum number of trials of socket binding, by sequentially
+  *                      increasing the port number.
+  * @param tcpBindTimeout Timeout for the tracker to bind to an InetSocketAddress.
+  */
+class RabitTracker private[scala](numWorkers: Int, port: Option[Int] = None,
+                                  maxPortTrials: Int = 1000, tcpBindTimeout: Duration = 1 minute)
+  extends IRabitTracker {
+
+  import scala.collection.JavaConverters._
+
+  require(numWorkers >=1, "numWorkers must be greater than or equal to one (1).")
+
+  val system = ActorSystem.create("RabitTracker")
+  val handler = system.actorOf(RabitTrackerHandler.props(numWorkers), "Handler")
+  implicit val askTimeout: akka.util.Timeout = akka.util.Timeout(30 seconds)
+
+  val futureWorkerEnvs: Future[Map[String, String]] = Try(
```

Comment:
This is because when the tracker sends `StartTracker` message to the actor, the actor has its internal messaging logic to handle Tcp port binding etc, which is async. 

Only when the port binding succeeds, the tracker can start properly. Therefore the future is used to behold to the "future" event of port binding.